File: sesion4-lab/aspnet-xxe/xxe-demo/Controllers/ProfileController.cs
using Microsoft.AspNetCore.Mvc;
using XxeDemo.Services;

namespace XxeDemo.Controllers;

/// <summary>
///  POST /api/profile/vulnerable     -> DtdProcessing.Parse (XXE explotable)
///  POST /api/profile/seguro         -> XmlDocument + XmlReaderSettings seguro
///  POST /api/profile/seguro-reader  -> solo XmlReader endurecido (diapositiva)
/// </summary>
[ApiController]
[Route("api/profile")]
public class ProfileController : ControllerBase
{
    private delegate XmlProfileParser.ParseResult XmlParser(string xml);

    [HttpPost("vulnerable")]
    [Consumes("application/xml")]
    public IActionResult Vulnerable() =>
        Parse(Request.Body, "VULNERABLE (DtdProcessing.Parse + XmlUrlResolver)", XmlProfileParser.ParseVulnerable);

    [HttpPost("seguro")]
    [Consumes("application/xml")]
    public IActionResult Seguro() =>
        Parse(Request.Body, "SEGURO (XmlDocument + XxeMitigationExample)", XmlProfileParser.ParseSeguro);

    [HttpPost("seguro-reader")]
    [Consumes("application/xml")]
    public IActionResult SeguroReader() =>
        Parse(Request.Body, "SEGURO (XmlReader endurecido — XxeMitigationExample)", XmlProfileParser.ParseSeguroReader);

    private IActionResult Parse(Stream body, string modo, XmlParser parser)
    {
        using var reader = new StreamReader(body);
        var xml = reader.ReadToEnd();

        var respuesta = new Dictionary<string, object?>
        {
            ["modo"] = modo,
            ["xmlRecibido"] = xml
        };

        try
        {
            var resultado = parser(xml);
            respuesta["usernameExtraido"] = resultado.Username;
            respuesta["exito"] = true;
        }
        catch (Exception ex)
        {
            respuesta["exito"] = false;
            respuesta["error"] = $"{ex.GetType().Name}: {ex.Message}";
        }

        return Ok(respuesta);
    }
}

File: sesion4-lab/spring-xxe/xxe-server/src/main/java/com/example/xxe/web/ProfileController.java
package com.example.xxe.web;

import java.util.LinkedHashMap;
import java.util.Map;

import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.example.xxe.xml.XmlProfileParser;
import com.example.xxe.xml.XmlProfileParser.ParseResult;

/**
 * Recibe perfiles de usuario en XML:
 *
 *  POST /api/profile/vulnerable   -> sin restricciones (XXE explotable)
 *  POST /api/profile/seguro       -> DOM con XxeMitigationExample
 *  POST /api/profile/seguro-sax   -> SAX con XxeMitigationExample
 */
@RestController
@RequestMapping("/api/profile")
public class ProfileController {

    @FunctionalInterface
    private interface XmlParser {
        ParseResult parse(String xml) throws Exception;
    }

    @PostMapping(value = "/vulnerable", consumes = MediaType.APPLICATION_XML_VALUE)
    public Map<String, Object> vulnerable(@RequestBody String xml) {
        return aRespuesta(
                "VULNERABLE (DocumentBuilderFactory sin restricciones)",
                xml,
                XmlProfileParser::parseVulnerable);
    }

    @PostMapping(value = "/seguro", consumes = MediaType.APPLICATION_XML_VALUE)
    public Map<String, Object> seguro(@RequestBody String xml) {
        return aRespuesta(
                "SEGURO (DocumentBuilderFactory endurecido — XxeMitigationExample)",
                xml,
                XmlProfileParser::parseSeguro);
    }

    @PostMapping(value = "/seguro-sax", consumes = MediaType.APPLICATION_XML_VALUE)
    public Map<String, Object> seguroSax(@RequestBody String xml) {
        return aRespuesta(
                "SEGURO (SAXParserFactory endurecido — XxeMitigationExample)",
                xml,
                XmlProfileParser::parseSeguroSax);
    }

    private Map<String, Object> aRespuesta(String modo, String xml, XmlParser parser) {
        Map<String, Object> respuesta = new LinkedHashMap<>();
        respuesta.put("modo", modo);
        respuesta.put("xmlRecibido", xml);

        try {
            ParseResult resultado = parser.parse(xml);
            respuesta.put("usernameExtraido", resultado.username());
            respuesta.put("exito", true);
        } catch (Exception e) {
            respuesta.put("exito", false);
            respuesta.put("error", e.getClass().getSimpleName() + ": " + e.getMessage());
        }

        return respuesta;
    }
}
